### PR TITLE
Modification of edge mask computation when l_fixed_area=T in horizontal remapping

### PR DESCRIPTION
**Pull request by JFLemieux73**
## PR checklist
- [x] Short (1 sentence) summary of your PR: 
   gx3, gx1 C grid simulations were previously crashing with horizontal remapping and  l_fixed_area=T. This modification corrects this problem. See issue #809 for details.
- [x] Developer(s): 
    @JFLemieux73 
- [x] Suggest PR reviewers from list in the column to the right.
    @eclare108213 
- [x] Please copy the PR test results link or provide a summary of testing completed below.

349 measured results of 349 total results
349 of 349 tests PASSED
0 of 349 tests PENDING
0 of 349 tests MISSING data
0 of 349 tests FAILED

- How much do the PR code changes differ from the unmodified code? 
    - [x] bit for bit
    - [ ] different at roundoff level
    - [ ] more substantial 
- Does this PR create or have dependencies on Icepack or any other models?
    - [ ] Yes
    - [x] No
- Does this PR update the Icepack submodule?  If so, the Icepack submodule must point to a hash on Icepack's main branch.
    - [ ] Yes
    - [x] No
- Does this PR add any new test cases?
    - [ ] Yes
    - [x] No
    Another PR will follow and will add a test case. I will also modify the code so that l_fixed_area=T when the C grid (and remap) is used. See issue #832.
- Is the documentation being updated? ("Documentation" includes information on the wiki or in the .rst files from doc/source/, which are used to create the online technical docs at https://readthedocs.org/projects/cice-consortium-cice/. A test build of the technical docs will be performed as part of the PR testing.)
    - [ ] Yes
    - [x] No, does the documentation need to be updated at a later time?
        - [x] Yes
        - [ ] No 
- [x] Please provide any additional information or relevant details below:

### CODE REVIEW

**Comment by eclare108213** [comment]
Looks good!  I suggest that you go ahead and add a little documentation (a couple of sentences?) and a test case to this PR.

**Comment by JFLemieux73** [comment]
Could you please merge this PR? I will add a test case and modifications to the doc in another PR as I would also like to make more modifications to the code so that l_fixed_area is set to true when the C grid is used. I have created an issue not to forget about this...

https://github.com/CICE-Consortium/CICE/issues/832

**Review by TillRasmussen — APPROVED** [review]
Looks fine.
One note. I suppose it is not bit for bit as it does not crash.

**Review by apcraig — COMMENTED** [review]
I would prefer if the do j, do i loops were aligned with respect to indentation like the other double nested loops in this subroutine.  But a minor quibble.

**Comment by apcraig** [comment]
What test suite was run?  I should probably run a full test suite including all the C-grid tests when ready.

Also, should a default change?  Should the C-grid always be running with l_fixed_area=T?

**Comment by JFLemieux73** [comment]
Hi Tony,

I ran the end-to-end testing procedure described in section 3.3.6.4 of the doc:

https://cice-consortium-cice.readthedocs.io/en/main/user_guide/ug_testing.html#end-to-end-testing-procedure

Yes l_fixed_area should always be true when using the C grid and I guess we should keep it set to false (default) for the B grid. I was not sure what is the best way to do that. I don't think I would add it in ice_in.

**Comment by apcraig** [comment]
You can add the namelist to

configuration/scripts/options/set_nml.gridc

when the C grid is set with cice.setup, it'll always then have the settings in set_nml.gridc.  If someone manually changes a case and turns on the C grid, then it's up to them to set l_fixed_area.  This should also probably be documented with the C grid.

**Comment by TillRasmussen** [comment]
> Hi Tony,
> 
> I ran the end-to-end testing procedure described in section 3.3.6.4 of the doc:
> 
> https://cice-consortium-cice.readthedocs.io/en/main/user_guide/ug_testing.html#end-to-end-testing-procedure
> 
> Yes l_fixed_area should always be true when using the C grid and I guess we should keep it set to false (default) for the B grid. I was not sure what is the best way to do that. I don't think I would add it in ice_in.

If this is the case then it should not be a hard coded parameter. Instead something like if c grid equal to true else if B grid equal to false else (cd grid I suppose) true/false

**Comment by JFLemieux73** [comment]
I am done with the modifications to the code and the doc (for l_fixed_area and C-grid). Let me know if these are ok.

**Comment by apcraig** [comment]
I'll run a full regression test suite on cheyenne including C-grid tests.  More soon.

**Comment by apcraig** [comment]
I tested the modifications on cheyenne.  Mostly all OK.  The cgrid results are largely not bit-for-bit, as expected.  Otherwise, things are bit-for-bit.  The only problem is that a few c and cd tests are failing that were not before,

```
FAIL cheyenne_intel_smoke_gbox80_11x3_boxslotcyl_gridc_reprosum_run10day run -1 -1 -1
FAIL cheyenne_intel_smoke_gbox80_8x1_boxslotcyl_cmplogrest_gridc_reprosum_run10day_thread run -1 -1 -1
FAIL cheyenne_intel_smoke_gbox80_1x1_boxslotcyl_gridc run -1 -1 -1
```

The error looks like 

```
 New mass < 0, i, j =           9           9
 Old mass =   1.00000000000000     
 New mass = -0.530255960863720     
 Net transport =  -1.53025596086372     
 (print_state) negative area (open
 (print_state) istep1, my_task, i, j, iblk:           1           0           9
           9          10
 ...
 (abort_ice) error = (diagnostic_abort)ERROR: negative area (open water)
```

We probably need to understand this a bit better before merging.

**Comment by apcraig** [comment]
Also, the full test results are here,

https://github.com/CICE-Consortium/Test-Results/wiki/cice_by_hash_forks#3d26d4cdb6089467b3c3a0866cb8f99fdc103f07

**Comment by JFLemieux73** [comment]
Thanks Tony. This is bad news...this is the problem we were trying to fix. So basically the new code fails for these test cases while the current main code works? 

I tried the regular box tests and gx3 configuration before the PR but I never tried the boxslotcyl...

**Comment by JFLemieux73** [comment]
I can reproduce the failure. I notice that for this test case: kdyn=-1. I just created an issue to investigate this:

https://github.com/CICE-Consortium/CICE/issues/836

**Comment by eclare108213** [comment]
> I can reproduce the failure. I notice that for this test case: kdyn=-1.

If kdyn=-1, then the dynamics should be turned off and ice velocities ought to be zero, unless they are being initialized to nonzero values for some reason.  Is this a problem with the test configuration rather than the code?

**Comment by JFLemieux73** [comment]
I don't know yet but for sure this is very different than the normal use of the model: velocities are initialized everywhere (even where there is no ice) and remain constant (in time) during the simulation.

**Comment by JFLemieux73** [comment]
I think this is not really an issue and that the boxsltcyl should be modified. The failure happens where there is no ice...In all the other idealized test cases or more realistic ones (gx3, gx1) the transport would not be calculated because aiu<amin. I coded something quickly just for testing and set the departure points to be zero (same as having uvel=vvel=0) when aiu is lower than amin (1e-03).  It does not crash anymore. Here is the hi field after 13 days (288 time steps = default for boxsltcyl) for the C grid simulation (with l_fixed_area=T):

CT_hi_2005-01-13-00000

As a reference here it is for the B grid simulation (l_fixed_area=F):

BF_hi_2005-01-13-00000

**Comment by JFLemieux73** [comment]
Maybe the C grid with l_fixed_area=T is more prone to these negative mass problems when there is no ice....but do we really want to investigate this further as this should never happen in normal simulations?

**Comment by apcraig** [comment]
is there a simple fix for the boxslotcyl test on the C-grid?  Change initialization or something else simple that won't affect the solver/solutions in general?

**Comment by eclare108213** [comment]
I think we need to look more closely at this test case, but I suspect that what it was intended to do is test the advection using constant ice velocities with the dynamics turned off.  The slotted cylinder ought to move in a predictable way, based on the ice velocities.  So if B grid is working and C grid isn't, then the masking must be acting differently?

**Comment by JFLemieux73** [comment]
Ok I think I understand what is happening. It is not a problem with remapping nor l_fixed_area. 

There are two things that need to be modified in the code. 

First with this test case some velocities are defined on land. The code should use umask and set uvel,vvel in set_state_var as:

if (umask(i,j)) then
                  uvel(i,j) =  c2*max_vel*(real(jglob(j), kind=dbl_kind) - p5) &
                            / real(ny_global - 1, kind=dbl_kind) - max_vel
                  vvel(i,j) = -c2*max_vel*(real(iglob(i), kind=dbl_kind) - p5) &
                            / real(nx_global - 1, kind=dbl_kind) + max_vel
else
                  uvel(i,j) = c0
                  vvel(i,j) = c0
endif

By doing this uvel(79,79) = 0 which is consistent with the landmask. uvel(78,78) is initialized to 2.27. As uvel(78,79) = 0 (no slip), uvelE(78,79) should be close to 1.135. uvelE(78,79) is however 2.27. To fix this, call grid_average_X2Y('S',uvel,'U',uvelE,'E')...in init_state should be replaced by call grid_average_X2Y('A',uvel,'U',uvelE,'E'). Same idea for the other grid_average_X2Y. 

I tested this and there is no more crash. The test case runs fine for both C grid (l_fixed_area=T) and B grid (l_fixed_area=F).

**Comment by apcraig** [comment]
Looking at ice_init, the code in question is,

 ```
     call bound_state (aicen,        &
                        vicen, vsnon, &
                        ntrcr, trcrn)

      if (grid_ice == 'CD' .or. grid_ice == 'C') then

         call grid_average_X2Y('S',uvel,'U',uvelN,'N')
         call grid_average_X2Y('S',vvel,'U',vvelN,'N')
         call grid_average_X2Y('S',uvel,'U',uvelE,'E')
         call grid_average_X2Y('S',vvel,'U',vvelE,'E')

         ! Halo update on North, East faces                                                                                                                   
         call ice_HaloUpdate(uvelN, halo_info, &
                             field_loc_Nface, field_type_scalar)
         call ice_HaloUpdate(vvelN, halo_info, &
                             field_loc_Nface, field_type_scalar)

         call ice_HaloUpdate(uvelE, halo_info, &
                             field_loc_Eface, field_type_scalar)
         call ice_HaloUpdate(vvelE, halo_info, &
                             field_loc_Eface, field_type_scalar)

      endif
```

This will only affect the initial condition of the C and CD configurations.  I assume it will change all answers for C and CD and make them ALL more correct?

**Comment by JFLemieux73** [comment]
Am I right that boxslotcyl is the only test case that has non zero initial velocities?

**Comment by apcraig** [comment]
> Am I right that boxslotcyl is the only test case that has non zero initial velocities?

I think that is true right now.  See the end of set_state_var in ice_init.F90.  So we should be OK for other test cases.  I will test again when updates are complete.

**Comment by JFLemieux73** [comment]
I thought I had written a comment here but I can't find it anymore...

Please do not merge this code yet. The model ran fine for 4 years with gx3. But I decided to also run gx1...unfortunately there was a crash after 10 months of simulation. I am investigating. See issue #809 for details.

**Comment by JFLemieux73** [comment]
Hi everyone,

Please have a look at issue #809. To solve the new failure with l_fixed_area=T will require more changes to the code. These changes will change answers on non-uniform grids (even for the B-grid...). I propose that we merge this PR first. I could then modify the code and run the QA testing to make sure everything is ok.

**Comment by apcraig** [comment]
I agree that a separate PR to modify the calculations for areafact makes sense.   I have read thru a bit of #809 and that plan all sounds reasonable.  If it's relatively easy to do the better solution ("solution 2"), then I think we should, but if it requires extra communication or storage at each timestep for additional grid metrics, then I think it can also be deferred.  I have also reviewed this PR and think it looks fine.  

Just to confirm, is this PR complete?  If so, I will run a quick test suite on cheyenne just to confirm everything is working then if others agree, we can merge and then @JFLemieux73 can create the next PR.

**Comment by JFLemieux73** [comment]
Thanks Tony. Yes this PR is complete.

File: cicecore/cicedyn/general/ice_init.F90
Comment (eclare108213) [inline cicecore/cicedyn/general/ice_init.F90:2748]: Why is this change in the averaging type needed?

**Review by eclare108213 — CHANGES_REQUESTED** [review]
The plan to merge this PR prior to making more invasive changes makes sense to me.  However, since the code will still not be working robustly, I think the documentation changes here should be reverted to the previous version.

**Comment by JFLemieux73** [comment]
To make sure it is more robust I should also change this in transport_remap:

if (grid_ice == 'CD' .or. grid_ice == 'C') then
         l_fixed_area = .true.
      else
         l_fixed_area = .false.
      endif

It should  be for the moment:

if (grid_ice == 'CD' .or. grid_ice == 'C') then
         l_fixed_area = .false.
      else
         l_fixed_area = .false.
      endif

Do you agree?

File: cicecore/cicedyn/general/ice_init.F90
Comment (JFLemieux73) [reply cicecore/cicedyn/general/ice_init.F90:2748]: This is just for the initialization of boxslotcyl test. uvelE and vvelN are defined from the uvel, vvel values. uvel(78,79) is initialized to 0 because of the land mask (no slip) while uvel(78,78) is initialized to 2.27. As uvelE is interpolated from uvel, uvelE(78,79) should be 2.27/2 =1.135. This is obtained by changing the averaging type.

**Comment by eclare108213** [comment]
I think you can leave l_fixed_area as true, since you're testing with it, but force other users to use upwind instead of remapping unless they manually change the code and so are aware of the issues.

File: cicecore/cicedyn/general/ice_init.F90
Comment (eclare108213) [reply cicecore/cicedyn/general/ice_init.F90:2748]: Ah, thank you.

**Comment by eclare108213** [comment]
Any changes to the defaults will affect @apcraig's tests, so decide what you want users to have while further work is done.

**Comment by JFLemieux73** [comment]
The current default is l_fixed_area = .false. for all the grids. It is simpler for the moment just to have it to false.  For the users we should just keep the warning that they should use upwind for the moment.

**Comment by JFLemieux73** [comment]
Ok this PR is ready. Let me know if there are any issues with the tests. Thanks!

**Comment by apcraig** [comment]
Before I restart testing, let me just ask a few questions.  

Why are we passing in "grid_ice" into init_remap and changing that interface?   We should be able to code "use ice_grid, only : grid_ice" in ice_transport_remap.F90, either in the module or subroutine init_remap.  This would simplify the changes.

If l_fixed_area is now a module variable in ice_transport_remap, then you could (should?) remove l_fixed_area from the locate_triangles subroutine since it's available as module data.

I assume the "if code" that defines l_fixed_area=.false. will be used in the next PR and made a bit more interesting?

Also, GHActions is currently broken, so we should ignore those failures for the time being.

@JFLemieux73, just confirm again if/when this is ready for testing.  Thanks!

**Comment by JFLemieux73** [comment]
Good points. I have removed l_fixed_area from the locate_triangles subroutine. I added use ice_grid, only : grid_ice in ice_transport_remap module. 

The if code is there for the next PR...

I think it is ready to be tested.

**Comment by apcraig** [comment]
Looks good, I'll start testing.

**Comment by apcraig** [comment]
Full test suite on cheyenne looks fine.  All tests pass and are bit-for-bit with current main except boxslotcyl which all pass but with different results.

7728 measured results of 7728 total results
7671 of 7728 tests PASSED
0 of 7728 tests PENDING
0 of 7728 tests MISSING data
57 of 7728 tests FAILED

**Review by apcraig — APPROVED** [review]

**Comment by apcraig** [comment]
Will merge in the morning unless anyone disagrees.

**Pull request merged**
